File: cpp/structure/linear/queue/QueueOfLinkedList.cpp
#include <iostream>
using namespace std;

class QueueOfLinkedList {
public:
    QueueOfLinkedList() : head(new ListNode(0)), tail(new ListNode(0)) {
        head->next = tail;
        tail->prev = head;
    }

    bool isEmpty() {
        return head->next == tail;
    }

    void enqueue(int value) {
        ListNode *node = new ListNode(value);
        node->prev = tail->prev;
        node->next = tail;
        tail->prev->next = node;
        tail->prev = node;
    }

    int dequeue() {
        if (head->next != tail) {
            ListNode *node = head->next;
            node->next->prev = head;
            head->next = node->next;
            return node->value;
        }
        return INT_MIN;
    }

    void clear() {
        head->next = tail;
        tail->prev = head;
    }

private:
    struct ListNode {
        int value;
        ListNode *prev;
        ListNode *next;

        ListNode(int value) : value(value), prev(nullptr), next(nullptr) {}
    };

    ListNode *head;
    ListNode *tail;
};

File: py/structure/linear/queue/QueueOfLinkedList.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class QueueOfLinkedList:
    def __init__(self):
        self.__head, self.__tail = self.__ListNode(0), self.__ListNode(0)
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    def isEmpty(self) -> bool:
        return self.__head.next == self.__tail

    def enqueue(self, value: int) -> None:
        node = self.__ListNode(value)
        node.prev, node.next, self.__tail.prev.next, self.__tail.prev = self.__tail.prev, self.__tail, node, node

    def dequeue(self) -> int:
        if self.__head.next != self.__tail:
            node = self.__head.next
            node.next.prev, self.__head.next = self.__head, node.next
            return node.value
        return -sys.maxsize

    def clear(self) -> None:
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    class __ListNode:
        def __init__(self, value: int):
            self.value, self.prev, self.next = value, None, None
